### ISSUE TITLE
CUBlas error when using O1, calling torch.matmul?

### ISSUE THREAD

**Issue by dave-epstein** [NVIDIA/apex#735]
Hi,

I have a loss function/training loop that works fine when using O2 but not when using O1. In particular, it throws the following error:

cublas runtime error : unknown error at /pytorch/aten/src/THC/THCBlas.cu:278

After the last line is executed in the following code fragment (variable names changed - there is more code afterwards that isn't reached):

```
a_reprs = reprs[flag== 0]
b_reprs= reprs[flag== 1]
a_vecs = a_reprs[:, 1:] - a_reprs[:, :-1]
a_vecs = a_vecs / (a_vecs.norm(p=2, dim=-1).unsqueeze(-1) + eps)
b_vecs = b_reprs[:, 1:] - b_reprs[:, :-1]
b_vecs = b_vecs / (b_vecs.norm(p=2, dim=-1).unsqueeze(-1) + eps)
a_cos = torch.triu(torch.matmul(a_vecs, a_vecs.transpose(-1, -2)), diagonal=1)
b_cos = torch.triu(torch.matmul(b_vecs,b_vecs.transpose(-1, -2)), diagonal=1)
```

I'm using the latest Apex version, CUDA 10.0, Pytorch 1.4.0, with Titan XP GPUs.

Glad to provide any other information that could be useful.

Thanks!

**Comment by ptrblck** [comment]
Could you post the shapes of the used tensors, so that we could reproduce this issue locally, please?

**Comment by dave-epstein** [comment]
Okay, so I actually figured this out, and I'm not sure if it's also a bug on your end or not, but the CUBlas issue was when the batch dimension of the matrices being multiplied was zero (in the code above, when flag was all zeroes or all ones). In O2 (and therefore in FP32), matmul/bmm did not crash when the batch dimension was zero, but it did in O1. I added a guard (`if any(flag == 0): ...`) which solved the problem, but maybe it is something relevant for you guys too.

Thanks for asking a good question which led me to figure the problem out :)
